### PR TITLE
[BugFix] Fix union all rewrite for str2date partitioned mv

### PR DESCRIPTION
**Pull request by LiShuMing**
Why I'm doing:

I created a mv like this and refreshed `2023-08-01`'s data.When I query it with `2023-08-01` and `2023-08-02` two days, I expected this query can be rewritten into mv by using union.

But it's not.
```
 select t1.a, t2.b, t1.d, bitmap_union(to_bitmap(t1.c))
 from  iceberg0.partitioned_db.part_tbl1 as t1 
 inner join iceberg0.partitioned_db.part_tbl2 t2 on t1.d=t2.d 
 inner join iceberg0.partitioned_db.part_tbl3 t3 on t1.d=t3.d 
 group by t1.a, t2.b, t1.d;
```
What I'm doing:
- Fix union bug: 
  -  For string type partition column, it should be list partition rather than range partition.
   - Only exclude `enforcedNonExistedColumns ` in the final rewritten plan which is not existed in `columnRefOperatorColumnMap ` otherwise will generate a wrong plan.
- Fix force rewrite for union: 
   - Add `appliedRuleMasks ` to mark this GroupExpression is derived from which rule.
   - For GroupExpression(which may contains UNION) which is drived from MULTI_MV rules, set it to zero totally rather than only MV node.
 - other minor bugfixs in debug. 

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/GroupExpression.java [hunk 1]
```diff
@@ -52,6 +52,7 @@ public class GroupExpression {
     private final List<Group> inputs;
     private final Operator op;
     private final BitSet ruleMasks = new BitSet(RuleType.NUM_RULES.ordinal() + 1);
+    private final BitSet appliedRuleMasks = new BitSet(RuleType.NUM_RULES.ordinal() + 1);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/GroupExpression.java:55]:
```
Add appliedRuleMasks  to mark this GroupExpression is derived from which rule.
```

why need this add appliedRuleMasks?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/GroupExpression.java:55]: Use this to track whether this GroupExpression is derived from MV rules or not. So then can be used in `ForceRewrite` later.

**Review by ABingHuang — DISMISSED** [review]

**Review by ABingHuang — DISMISSED** [review]

**Review by ABingHuang — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerTraceUtil.java [hunk 2]
```diff
@@ -86,14 +86,14 @@ public static void logApplyRule(OptimizerContext ctx, Rule rule,
         Tracers.log(Tracers.Module.OPTIMIZER, args -> {
             StringBuilder sb = new StringBuilder();
             sb.append(String.format("[TRACE QUERY %s] APPLY RULE %s\n", ctx.getQueryId(), rule));
-            sb.append("Original Expression:\n").append(oldExpression.debugString(3));
+            sb.append("Original Expression:\n").append(oldExpression.debugString());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerTraceUtil.java:89]: maybe vary large

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java [hunk 3]
```diff
@@ -383,4 +385,24 @@ public String debugString(String headlineIndent, String detailIndent) {
         return sb.toString();
     }
 
+    public boolean hasMVGroupExpression() {
+        return logicalExpressions.stream().anyMatch(x -> x.hasAppliedMVRules()) ||
+                physicalExpressions.stream().anyMatch(x -> x.hasAppliedMVRules());
+    }
+
+    public void invalidNonMVGroupExpressions(TaskContext taskContext) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Group.java:393]:
```suggestion
    public void forceChooseMVExpression(TaskContext taskContext) {
```

**Review by Seaven — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Pull request merged**